### ISSUE TITLE
Shape_inference can't infer shape for onnx.loop innergraph(loop body)

### ISSUE THREAD

**Issue by lanhongvp** [onnx/onnx#3312]
# Bug Report

### Is the issue related to model conversion?
No

### Describe the bug
`onnx.shape_inference.infer_shapes` can't infer shape for onnx.loop inner-graph(since onnx.loop attribute is a graph). See the screen shot for reference.

![loopbody in Netron](https://user-images.githubusercontent.com/30483478/109916474-ab201800-7cee-11eb-9fca-d73116f40dea.png)

### System information
- OS Platform and Distribution (*e.g. Linux Ubuntu 16.04*):  Windows.
- ONNX version (*e.g. 1.7*):  1.8.0
- Python version: 3.7.6
- GCC/Compiler version (if compiling from source): N/A
- CMake version: N/A
- Protobuf version: 3.14.0
- Visual Studio version (if applicable): Check thru VS code
- VS code version: 1.53

### Reproduction instructions

```
import torch
import onnx
import onnxruntime.tools.symbolic_shape_infer as sym
import numpy as np

@torch.jit.script
def loop(x, y):
    for i in range(int(y)):
        x = x + i
    return x

class LoopModel2(torch.nn.Module):
    def forward(self, x, y):
        return loop(x, y)

if __name__ == '__main__':
    model = LoopModel2()
    torch.onnx.export(model, (torch.tensor([[]]), torch.tensor([10], dtype=torch.long)), 'loop.onnx', verbose=True,
                    input_names = ['input_data', 'loop_range'],
                    output_names = ['result'],
                    dynamic_axes = { 'input_data': [0, 1] })
    model = onnx.load('loop.onnx')
    
    # override input shapes
    model.graph.input[0].type.tensor_type.shape.dim[0].dim_value = 3
    model.graph.input[0].type.tensor_type.shape.dim[1].dim_value = 5
    onnx.save(model, 'loop_override.onnx')

    # using ONNX shape inference    
    inferModel = onnx.shape_inference.infer_shapes(model)
    onnx.save(inferModel, 'loop_override_inferred.onnx')
    
    # using symbolic shape inference
    out_mp = sym.SymbolicShapeInference.infer_shapes(in_mp=model, auto_merge=True)
    onnx.save(out_mp, 'loop_override_symbolic.onnx')
```
- Attach the ONNX model to the issue (where applicable)

> - [Onnx shape inferred models with loop](https://github.com/lanhongvp/onnxModels/blob/main/onnxTestModels/dynamic_loop_override_inferred.onnx)

### Expected behavior
see the screen shot of onnx runtime `SymbolicShapeInference` . This is what we expect for `onnx.shape_inference` to handle the loopbody in onnx.loop.

![symoblic shape inferred model in Netron](https://user-images.githubusercontent.com/30483478/109916668-f9cdb200-7cee-11eb-8e94-3a324b3c0e99.png)

**Comment by gramalingam** [comment]
Yes, this is a limitation of the loop shape inference. The reason for this is that the shapes of loop state-variables can vary across different iterations (depending on the ops used inside the loop body). Of course, in some cases, like the one above, the shapes may remain the same across all iterations. But this requires some extra-analysis (we need to analyze the loop body more than once to reach a safe fixed-point). Hence, it is not done currently. But it would be useful to do it. PRs to improve the analysis welcome!
